File: backend/Api/WorkOrder/WorkOrder.Entity/Models/WorkOrderModels.cs
using System.ComponentModel.DataAnnotations;
using WorkOrder.Entity.Enums;

namespace WorkOrder.Entity.Models;

// Requests and views for the wrk API (S6, TK-66).

public sealed record WorkOrderMessage(string Message);

public sealed class SaveWorkOrderRequest
{
    [Required(ErrorMessage = "Title is required.")]
    [MaxLength(200, ErrorMessage = "Title cannot exceed 200 characters.")]
    public string Title { get; set; } = null!;

    [MaxLength(2000, ErrorMessage = "Description cannot exceed 2000 characters.")]
    public string? Description { get; set; }

    public WorkOrderSource WorkOrderSource { get; set; } = WorkOrderSource.Complaint;

    public WorkOrderPriority Priority { get; set; } = WorkOrderPriority.Medium;

    public long? FacilityAssetId { get; set; }

    public long? SpaceId { get; set; }

    public DateOnly ReportedDate { get; set; }

    public DateOnly? DueDate { get; set; }

    public List<string> Tasks { get; set; } = [];
}

public sealed class WorkOrderActionRequest
{
    public WorkOrderAction Action { get; set; }

    /// <summary>For Assign.</summary>
    public long? EmployeeId { get; set; }

    /// <summary>For Complete.</summary>
    public DateOnly? CompletedDate { get; set; }

    /// <summary>For Complete.</summary>
    [Range(typeof(decimal), "0", "100000000", ErrorMessage = "Labour cost cannot be negative.")]
    public decimal? LabourCost { get; set; }

    /// <summary>For Cancel.</summary>
    [MaxLength(500, ErrorMessage = "Reason cannot exceed 500 characters.")]
    public string? Reason { get; set; }
}

public sealed class TaskTickRequest
{
    public bool IsDone { get; set; }
}

public sealed class IssuePartRequest
{
    [Range(1, long.MaxValue, ErrorMessage = "Choose an item.")]
    public long ItemId { get; set; }

    public long? WarehouseId { get; set; }

    [Range(typeof(decimal), "0.0001", "1000000", ErrorMessage = "The quantity must be more than zero.")]
    public decimal Quantity { get; set; }

    public DateOnly IssueDate { get; set; }
}

/// <summary>What another service asks to raise (TK-67 plans, TK-68 AMC visits). Idempotent on <see cref="SourceKey"/>.</summary>
public sealed class RaiseWorkOrderRequest
{
    public Guid CustomerId { get; set; }

    public Guid OrgId { get; set; }

    [Required(ErrorMessage = "Source key is required.")]
    [MaxLength(60, ErrorMessage = "Source key cannot exceed 60 characters.")]
    public string SourceKey { get; set; } = null!;

    [Required(ErrorMessage = "Title is required.")]
    [MaxLength(200, ErrorMessage = "Title cannot exceed 200 characters.")]
    public string Title { get; set; } = null!;

    public WorkOrderSource WorkOrderSource { get; set; } = WorkOrderSource.Preventive;

    public long? FacilityAssetId { get; set; }

    public long? SpaceId { get; set; }

    public DateOnly ReportedDate { get; set; }

    public DateOnly? DueDate { get; set; }

    public long? AssignedEmployeeId { get; set; }

    public long? PreventivePlanId { get; set; }

    public long? AmcContractId { get; set; }
}

public sealed class RaiseWorkOrderResponse
{
    public long WorkOrderId { get; set; }

    public string WorkOrderNo { get; set; } = null!;

    /// <summary>False when one with that source key already existed.</summary>
    public bool Created { get; set; }
}

public sealed class WorkOrderTaskView
{
    public long WorkOrderTaskId { get; set; }

    public string Description { get; set; } = null!;

    public bool IsDone { get; set; }
}

public sealed class WorkOrderPartView
{
    public long WorkOrderPartId { get; set; }

    public long ItemId { get; set; }

    public string? ItemName { get; set; }

    public long? WarehouseId { get; set; }

    public decimal Quantity { get; set; }

    public decimal UnitCost { get; set; }
}

public sealed class WorkOrderView
{
    public long WorkOrderId { get; set; }

    public string WorkOrderNo { get; set; } = null!;

    public string Title { get; set; } = null!;

    public string? Description { get; set; }

    public WorkOrderSource WorkOrderSource { get; set; }

    public WorkOrderPriority Priority { get; set; }

    public long? FacilityAssetId { get; set; }

    public long? SpaceId { get; set; }

    public DateOnly ReportedDate { get; set; }

    public DateOnly? DueDate { get; set; }

    public long? AssignedEmployeeId { get; set; }

    public long? AmcContractId { get; set; }

    public long? PreventivePlanId { get; set; }

    public WorkOrderStatus WorkOrderStatus { get; set; }

    public DateOnly? CompletedDate { get; set; }

    public decimal LabourCost { get; set; }

    public decimal PartsCost { get; set; }

    public string? CancelReason { get; set; }

    public List<WorkOrderTaskView> Tasks { get; set; } = [];

    public List<WorkOrderPartView> Parts { get; set; } = [];
}

File: frontend/libs/work-order/work-order-core/src/lib/work-order.models.ts
/** The work order service's shapes (S6, TK-66), as the API sends them: enums by name. */

export type WorkOrderSource = 'Complaint' | 'Preventive' | 'Amc' | 'Inspection';
export type WorkOrderPriority = 'Low' | 'Medium' | 'High' | 'Urgent';
export type WorkOrderStatus = 'Open' | 'Assigned' | 'InProgress' | 'OnHold' | 'Completed' | 'Closed' | 'Cancelled';
export type WorkOrderAction = 'Assign' | 'Start' | 'Hold' | 'Resume' | 'Complete' | 'Close' | 'Cancel';

export interface WorkOrderTask {
  workOrderTaskId: number;
  description: string;
  isDone: boolean;
}

export interface WorkOrderPart {
  workOrderPartId: number;
  itemId: number;
  itemName: string | null;
  warehouseId: number | null;
  quantity: number;
  unitCost: number;
}

export interface WorkOrder {
  workOrderId: number;
  workOrderNo: string;
  title: string;
  description: string | null;
  workOrderSource: WorkOrderSource;
  priority: WorkOrderPriority;
  facilityAssetId: number | null;
  spaceId: number | null;
  reportedDate: string;
  dueDate: string | null;
  assignedEmployeeId: number | null;
  amcContractId: number | null;
  preventivePlanId: number | null;
  workOrderStatus: WorkOrderStatus;
  completedDate: string | null;
  labourCost: number;
  partsCost: number;
  cancelReason: string | null;
  tasks: WorkOrderTask[];
  parts: WorkOrderPart[];
}

export interface SaveWorkOrder {
  title: string;
  description: string | null;
  workOrderSource: WorkOrderSource;
  priority: WorkOrderPriority;
  facilityAssetId: number | null;
  spaceId: number | null;
  reportedDate: string;
  dueDate: string | null;
  tasks: string[];
}

export interface WorkOrderActionBody {
  action: WorkOrderAction;
  employeeId?: number | null;
  completedDate?: string | null;
  labourCost?: number | null;
  reason?: string | null;
}

export interface IssuePart {
  itemId: number;
  warehouseId: number | null;
  quantity: number;
  issueDate: string;
}

export interface StockItem {
  itemId: number;
  itemCode: string;
  itemName: string;
  quantityOnHand: number;
}

export interface StockWarehouse {
  warehouseId: number;
  warehouseCode: string;
  warehouseName: string;
}

export const WORK_ORDER_STATUSES: readonly { value: WorkOrderStatus; label: string }[] = [
  { value: 'Open', label: 'Open' },
  { value: 'Assigned', label: 'Assigned' },
  { value: 'InProgress', label: 'In progress' },
  { value: 'OnHold', label: 'On hold' },
  { value: 'Completed', label: 'Completed' },
  { value: 'Closed', label: 'Closed' },
  { value: 'Cancelled', label: 'Cancelled' },
];

export const WORK_ORDER_PRIORITIES: readonly { value: WorkOrderPriority; label: string }[] = [
  { value: 'Low', label: 'Low' },
  { value: 'Medium', label: 'Medium' },
  { value: 'High', label: 'High' },
  { value: 'Urgent', label: 'Urgent' },
];

export const WORK_ORDER_SOURCES: readonly { value: WorkOrderSource; label: string }[] = [
  { value: 'Complaint', label: 'Complaint' },
  { value: 'Preventive', label: 'Preventive plan' },
  { value: 'Amc', label: 'AMC visit' },
  { value: 'Inspection', label: 'Inspection' },
];

/**
 * The moves a work order offers from where it stands — the same table the
 * server enforces, so a screen never offers a button the server will refuse.
 * Close is left out: it has its own button and its own permission.
 */
export function nextActions(status: WorkOrderStatus): WorkOrderAction[] {
  switch (status) {
    case 'Open':
      return ['Assign', 'Cancel'];
    case 'Assigned':
      return ['Assign', 'Start', 'Cancel'];
    case 'InProgress':
      return ['Hold', 'Complete'];
    case 'OnHold':
      return ['Resume'];
    default:
      return [];
  }
}

/** Only an Open work order can be edited; parts are issued while the work is under way. */
export const isEditable = (status: WorkOrderStatus): boolean => status === 'Open';
export const takesParts = (status: WorkOrderStatus): boolean => status === 'Assigned' || status === 'InProgress' || status === 'OnHold';
